File: src/main/java/com/ratelimiter/RateLimiter.java
package com.ratelimiter;

import com.ratelimiter.strategy.RateLimitStrategy;
import com.ratelimiter.strategy.TokenBucketStrategy;

public class RateLimiter {

    public enum Algorithm {
        TOKEN_BUCKET
    }

    private final RateLimitStrategy strategy;

    public RateLimiter(Algorithm algorithm) {
        if (algorithm == Algorithm.TOKEN_BUCKET) {
            this.strategy = new TokenBucketStrategy();
        } else {
            throw new IllegalArgumentException("Unsupported algorithm: " + algorithm);
        }
    }

    public RateLimiter(RateLimitStrategy strategy) {
        this.strategy = strategy;
    }

    public RateLimitResponse hit(String identifier, RateLimitOptions options) {
        if (options == null || options.getLimit() <= 0 || options.getWindow() <= 0) {
            throw new IllegalArgumentException("Options 'limit' and 'window' are required and must be > 0.");
        }

        try {
            return strategy.hit(identifier, options);
        } catch (Exception e) {
            // Log error
            System.err.println("RateLimiter Error: " + e.getMessage());
            e.printStackTrace();

            // Fail-open: Permit the request if an internal exception occurs
            return new RateLimitResponse(
                    true,
                    Math.max(0, options.getLimit() - options.getCost()),
                    (System.currentTimeMillis() / 1000) + options.getWindow());
        }
    }
}

File: src/RateLimiter.js
const RedisClient = require('./RedisClient');
const TokenBucket = require('./strategies/TokenBucket');

class RateLimiter {
    constructor(config) {
        this.redis = RedisClient.connect(config.redisUrl || 'redis://localhost:6379');
        this.keyPrefix = config.keyPrefix || 'rl:';

        switch (config.algorithm) {
            case 'TOKEN_BUCKET':
                this.strategy = new TokenBucket(this.redis, this.keyPrefix);
                break;
            default:
                throw new Error(`Invalid algorithm: ${config.algorithm}`);
        }
    }

    /**
     * @param {string} identifier
     * @param {object} options
     * @returns {Promise<{allowed: boolean, remaining: number, resetTime: number}>}
     */
    async hit(identifier, options) {
        if (!options || !options.limit || !options.window) {
            throw new Error('Options "limit" and "window" are required.');
        }

        try {
            return await this.strategy.hit(identifier, options);
        } catch (err) {
            console.error('RateLimiter Error:', err);
            // Fail-open
            return {
                allowed: true,
                remaining: options.limit - 1,
                resetTime: Date.now() / 1000 + options.window
            };
        }
    }
}

module.exports = RateLimiter;
